File: benchmarks/C++_Primitives/TrafficLight.cpp
#include <iostream>
#include <string>
#include <btypes_primitives/BUtils.hpp>
#include <btypes_primitives/BSet.hpp>
#include <btypes_primitives/BObject.hpp>
#include <btypes_primitives/BBoolean.hpp>

#ifndef TrafficLight_H
#define TrafficLight_H

using namespace std;

class TrafficLight {

    public:
        class colors : public BObject {
            public:

                typedef void value_type;
                typedef void left_type;
                typedef void right_type;

                enum colors_type {
                    red, 
                    redyellow, 
                    yellow, 
                    green
                };

                colors_type value;

                colors(){}

                colors(const colors_type& type) {
                    this->value = type;
                }

                BBoolean equal(const colors& o) {
                    return value == o.value;
                }

                BBoolean unequal(const colors& o) {
                    return value != o.value;
                }

                friend bool operator ==(const colors& p1, const colors& p2) {
                    return p1.value == p2.value;
                }

                friend bool operator !=(const colors& p1, const colors& p2) {
                    return p1.value != p2.value;
                }

                void operator =(const colors& other) {
                    this->value = other.value;
                }

                friend std::ostream& operator<<(std::ostream &strm, const colors& e) {
                    switch(e.value) {
                        case red: return strm << "red";
                        case redyellow: return strm << "redyellow";
                        case yellow: return strm << "yellow";
                        case green: return strm << "green";
                    }
                }

                int hashCode() const {
                    return static_cast<int>(value);
                }
        };

    private:



        #define _colors (BSet<colors >((colors(colors::red)), (colors(colors::redyellow)), (colors(colors::yellow)), (colors(colors::green))))

        colors tl_cars;
        colors tl_peds;

    public:

        TrafficLight() {
            tl_cars = (colors(colors::red));
            tl_peds = (colors(colors::red));
        }

        void cars_ry() {
            if((tl_cars.equal((colors(colors::red)))._and(tl_peds.equal((colors(colors::red))))).booleanValue()) {
                tl_cars = (colors(colors::redyellow));
            }
        }

        void cars_y() {
            if((tl_cars.equal((colors(colors::green)))._and(tl_peds.equal((colors(colors::red))))).booleanValue()) {
                tl_cars = (colors(colors::yellow));
            }
        }

        void cars_g() {
            if((tl_cars.equal((colors(colors::redyellow)))._and(tl_peds.equal((colors(colors::red))))).booleanValue()) {
                tl_cars = (colors(colors::green));
            }
        }

        void cars_r() {
            if((tl_cars.equal((colors(colors::yellow)))._and(tl_peds.equal((colors(colors::red))))).booleanValue()) {
                tl_cars = (colors(colors::red));
            }
        }

        void peds_r() {
            if((tl_peds.equal((colors(colors::green)))._and(tl_cars.equal((colors(colors::red))))).booleanValue()) {
                tl_peds = (colors(colors::red));
            }
        }

        void peds_g() {
            if((tl_peds.equal((colors(colors::red)))._and(tl_cars.equal((colors(colors::red))))).booleanValue()) {
                tl_peds = (colors(colors::green));
            }
        }

};

#endif

File: benchmarks/execution/Rust_embedded/TrafficLight.rs
#![ allow( dead_code, unused, non_snake_case, non_camel_case_types, unused_assignments ) ]
//#![no_std] //std is enabled for the generated machine by default, since without it there is no println which makes unit-testing very difficult
use btypes::{bset, brel};
use btypes::bset::BSet;
use btypes::bset::SetItem;
use btypes::bset::PowSetItem;
use btypes::bset::PowAble;
use btypes::bset::NestedSet;
use btypes::brelation::BRelation;
use btypes::brelation::RelLeftItem;
use btypes::brelation::RelPowAble;
use btypes::bboolean::BBoolean;
use btypes::bboolean::BBool;
use btypes::bboolean::BOOL;
use btypes::binteger::BInteger;
use btypes::binteger::BInt;
use btypes::binteger::set_BInteger;
use btypes::btuple::BTuple;




//set_enum_declaration
#[derive(Default, Debug, Clone, Copy, PartialEq, Eq, Hash, PartialOrd, Ord)]
pub enum colors {
    #[default]
    red = 0, 
    redyellow = 1, 
    yellow = 2, 
    green = 3
}

impl colors {
    pub fn equal(&self, other: &colors) -> bool { *self == *other }
    pub fn unequal(&self, other: &colors) -> bool { *self != *other }
    pub const fn to_idx(self) -> usize { self as usize }
    pub fn from_idx(idx: usize) -> Self {
        match idx {
            0 => Self::red, 
            1 => Self::redyellow, 
            2 => Self::yellow, 
            3 => Self::green, 
            _ => panic!("colors index out of range! {:?}", idx)
        }
    }
}

impl SetItem<4> for colors {
    fn as_idx(&self) -> usize { self.to_idx() }
    fn from_idx(idx: usize) -> Self { Self::from_idx(idx) }
}

type set_colors = BSet<colors, 4>;
//set_enum_declaration done


#[derive(Default, Debug)]pub struct TrafficLight {
    /*declarations*/tl_cars: colors,
    tl_peds: colors,
    /*sets*//*set_declaration*/_colors: set_colors,
}


impl TrafficLight {

    pub fn new() -> TrafficLight {
        //values: ''
        let mut m: TrafficLight = Default::default();
        m.init();
        return m;
    }
    fn init(&mut self) {
        /*set_initializations*///set_initialization
    self._colors = bset![colors, /*enum_call*/colors::red, /*enum_call*/colors::redyellow, /*enum_call*/colors::yellow, /*enum_call*/colors::green];
        /*properties*/
        /*body*/self.tl_cars = /*enum_call*/colors::red;
    self.tl_peds = /*enum_call*/colors::red;
        /*includesInitialization*/
    }

    pub fn get_tl_cars(&self) -> colors {
        return self.tl_cars.clone();
    }

    pub fn get_tl_peds(&self) -> colors {
        return self.tl_peds.clone();
    }

    pub fn get__colors(&self) -> set_colors {
        return self._colors.clone();
    }

    pub fn cars_ry(&mut self) -> () {
        //select
        if (self.tl_cars.equal(&/*enum_call*/colors::red) && self.tl_peds.equal(&/*enum_call*/colors::red)) {
            self.tl_cars = /*enum_call*/colors::redyellow;
        } else {
            panic!("ERROR: called SELECT-function with incompatible parameters!");
        }
    }

    pub fn cars_y(&mut self) -> () {
        //select
        if (self.tl_cars.equal(&/*enum_call*/colors::green) && self.tl_peds.equal(&/*enum_call*/colors::red)) {
            self.tl_cars = /*enum_call*/colors::yellow;
        } else {
            panic!("ERROR: called SELECT-function with incompatible parameters!");
        }
    }

    pub fn cars_g(&mut self) -> () {
        //select
        if (self.tl_cars.equal(&/*enum_call*/colors::redyellow) && self.tl_peds.equal(&/*enum_call*/colors::red)) {
            self.tl_cars = /*enum_call*/colors::green;
        } else {
            panic!("ERROR: called SELECT-function with incompatible parameters!");
        }
    }

    pub fn cars_r(&mut self) -> () {
        //select
        if (self.tl_cars.equal(&/*enum_call*/colors::yellow) && self.tl_peds.equal(&/*enum_call*/colors::red)) {
            self.tl_cars = /*enum_call*/colors::red;
        } else {
            panic!("ERROR: called SELECT-function with incompatible parameters!");
        }
    }

    pub fn peds_r(&mut self) -> () {
        //select
        if (self.tl_peds.equal(&/*enum_call*/colors::green) && self.tl_cars.equal(&/*enum_call*/colors::red)) {
            self.tl_peds = /*enum_call*/colors::red;
        } else {
            panic!("ERROR: called SELECT-function with incompatible parameters!");
        }
    }

    pub fn peds_g(&mut self) -> () {
        //select
        if (self.tl_peds.equal(&/*enum_call*/colors::red) && self.tl_cars.equal(&/*enum_call*/colors::red)) {
            self.tl_peds = /*enum_call*/colors::green;
        } else {
            panic!("ERROR: called SELECT-function with incompatible parameters!");
        }
    }
}


